Moment js to date-fns migration

srijan-paul opened the pull request:
This PR introduces a new migration for `moment` -> `date-fns`.
Right now, nearly all the core functionality has been migrated.

The only caveats that remain are:
1. Plugins (don't plan on supporting these)
2. Formatting / Parsing spec (Work in progress)
3. [Customizing the library's behavior itself](https://momentjs.com/docs/#/customization/) (this cannot be ported since date-fns does not support custom locales the way moment does. I suggest removing all calls that modify the locale state globally).
4. mutation isn't fully supported. Migrations like these won't be fully accurate:
```js
foo.bar.setQuarter(q);
// becomes:
/* TODO: moment.js dates are mutable – feed this value back through properly */
datefns.setQuarter(foo.bar, q)
```

morgante commented:
> All method calls on a Duration object can be rewritten to a method call from one of these libraries, but I wonder if introducing another package is fine?

Yes, let's simply use the other library when necessary.

> Is there a way to define a function in an external module (ideally), or inside the current file, that can be referenced in the pattern?

You can always insert new code into the current `$program` by finding a node and appending content after that. That being said, I would probably lean towards using an IIFE in these cases - it's more bloated, but easier to understand.

> Since a format specifier cannot always be determined statically, we should also generate a function that transforms a moment-js specifier to a date-fns specifier.

Yes, let's use an IIFE *if necessary*. We should still try to do it statically when we can.

> moment-js has several of official (and third-party) [plugins](https://momentjs.com/docs/#/plugins/).

We can leave plugins out of scope.

morgante reviewed (commented):
Just a note that we strongly prefer TDD - you should move this to a `.md` file with samples once you are ready.

morgante commented on .grit/patterns/moment-to-date-fns.grit:
```diff
@@ -0,0 +1,219 @@
+// A call to `add`/`sub` method of a `momentjs` date object.
+pattern moment_arith($date, $method, $count, $pat) {
+    `$date.$method($count, $pat)` as $arith_exp where {
+        or { $method <: "add", $method <: "sub" },
```
```suggestion
        $method <: or { `add`, `sub` },
```

morgante commented on .grit/patterns/moment-to-date-fns.grit:
```diff
@@ -0,0 +1,219 @@
+pattern moment_arith($date, $method, $count, $pat) {
+    `$date.$method($count, $pat)` as $arith_exp where {
+        or { $method <: "add", $method <: "sub" },
+        if ($method <: "add") {
```
Use snippets for code. Not quotes.

```suggestion
        if ($method <: `add`) {
```

srijan-paul commented:
@morgante The [documentation](https://docs.grit.io/language/patterns#rewrite-operator) for patterns mentions that metavariable names can be wrapped in `{}` to avoid conflating them with alphanumeric characters that follow.
This snippet is given in the examples:
<img width="715" alt="Screenshot 2023-09-27 at 1 30 34 PM" src="https://github.com/getgrit/js/assets/50487716/80c8e360-fcc1-4900-a401-b186b48b015e">

However, the same code snippet does not seem to work in the playground:

<img width="864" alt="Screenshot 2023-09-27 at 1 31 37 PM" src="https://github.com/getgrit/js/assets/50487716/c43122bb-1362-4179-9741-d3771f353ed9">

Currently, the only way to interpolate expressions seems to be the `join` builtin.
Is this a bug or am I missing something?

morgante commented:
> @morgante The [documentation](https://docs.grit.io/language/patterns#rewrite-operator) for patterns mentions that metavariable names can be wrapped in `{}` to avoid conflating them with alphanumeric characters that follow. This snippet is given in the examples: <img alt="Screenshot 2023-09-27 at 1 30 34 PM" width="715" src="https://user-images.githubusercontent.com/50487716/270906549-80c8e360-fcc1-4900-a401-b186b48b015e.png">
> 
> However, the same code snippet does not seem to work in the playground:
> 
> <img alt="Screenshot 2023-09-27 at 1 31 37 PM" width="864" src="https://user-images.githubusercontent.com/50487716/270907160-c43122bb-1362-4179-9741-d3771f353ed9.png">
> Currently, the only way to interpolate expressions seems to be the `join` builtin. Is this a bug in the compiler or am I missing something?

It's a bug / partial regression. We are supposed to support that syntax for right-hand side snippets but not for left-hand side ones. For now just use join syntax.

srijan-paul commented:
All Get+Set, Querying, Manipulation functions are more or less covered.
Remaining: i18n, parse and display, complete support for durations (currently half baked).

morgante commented on .grit/patterns/moment-to-date-fns.grit:
```diff
@@ -0,0 +1,358 @@
+   `$duration_or_date.$accessor($arg)` where {
+    $accessor <: r"(year|month|week|day|hour|minute|second|millisecond)s?"($unit),
+    $capitalized_unit = capitalize(string=$unit),
+    $accessor_fn = join(list=["get", $capitalized_unit, "s"], separator=""),
```
@srijan-paul We pushed an update so you can use braces: https://docs.grit.io/language/patterns#metavariables

```suggestion
    $accessor_fn = `get$[capitalized_unit]s`
```

srijan-paul commented:
@morgante Is there some way to ensure default imports using `ensure_import_from`?
Currently, it seems to only perform non-default imports.

e.g: `$foo <: ensure_import_from(source="foo")` will do:

```js
import { foo } from "foo"
```

Whereas I want:

```js
import foo from "foo"
// or:
import * as foo from "foo"
```

I ask because not all node-js packages expose themselves as ES Modules.

srijan-paul commented:
I noticed an odd behavior with `sequential` that (I think) isn't documented.
It is supposed to run multiple patterns in order, but when one of the patterns in a sequence fails to match any node, none of the following patterns are run.

Is this behavior intentional?
If so, is there some way to run multiple patterns in sequence even if one doesn't match any node?
```
sequential {
  replace_imports(),
  replace_expressions() // if there are no imports, this is never run.
}
```

srijan-paul commented:
@morgante I think this PR could use a round of review now.
I'll un-draft it as soon as I've figured out the formatting bit, but the rest of the library has been ported fully.

I've added some unit tests, though the coverage isn't 100% yet.
Will follow up with more tests.

morgante commented on .grit/patterns/moment-to-date-fns.grit:
```diff
@@ -0,0 +1,765 @@
+
+    `$date.$func($arg)` as $exp where {
+      $func <: or {
+        r"dates?" => `Month`,
+        r"(?:days?)|(?:weekdays?)" => `Day`,
```
Avoid regex when feasible.

```suggestion
       or { `date`, `dates` } => `Month`,
       or { `day`, `days`, `weekday`, `weekdays` } => `Day`,
```

(In theory we can add a compiler optimization for this. But for now this is better.)

morgante commented on .grit/patterns/moment-to-date-fns.grit:
```diff
@@ -0,0 +1,765 @@
+// This only affects the way a date is displayed (PST vs UTC).
+// This could technically be migrated by maintaining a global map of flags and
+// turning every `utc()` call into `set`, and every `display` call into a custom
+// overload, but I think that would introduce too much unwanted noise into the user's program.
```
Agreed, but let's leave a comment.

morgante commented on .grit/patterns/moment-to-date-fns.grit:
```diff
@@ -0,0 +1,765 @@
+
+    missing_datefns_features(),
+
+    `$normalize` where {
+      $normalize <: or {
+        `$_.normalizeUnits`,
+        `normalizeUnits`
+      }
+    } => `normalizeMomentJSUnit`,
```
```suggestion
    or {
        `$_.normalizeUnits`,
        `normalizeUnits`
      } => `normalizeMomentJSUnit`,
```

morgante commented on .grit/patterns/moment-to-date-fns.grit:
```diff
@@ -0,0 +1,765 @@
+      }
+    } => `normalizeMomentJSUnit`,
+
+    `$isValid` where {
+      $isValid <: or { `$_.isValid`, `isValid` },
+    } => `datefns.isValid`,
```
```suggestion
    or { `$_.isValid`, `isValid` } => `datefns.isValid`,
```

morgante commented on .grit/patterns/moment-to-date-fns.grit:
```diff
@@ -0,0 +1,765 @@
+  } => `$date`
+}
+
+pattern rewrite_misc_methods() {
```
Many of these are far too broad - ex. `$d.toJSON()` here will catch lots of *non-date* JSON objects. Let's make sure to tighten it up and add some negative test cases (code we shouldn't convert).

srijan-paul replied:
It'll be impossible to always accurately determine the correct dispatch for `d` statically.
Instead, I think I'll do:

```js
(d instanceof Date ? ... : d.toJSON())
```

This way, all method calls matching `$d.toJSON` will be re-written, but the behavior won't change.

morgante replied:
It's still too broad. Between rewriting all toJSON instances throughout an app and potentially missing a few moment instances I lean heavily towards missing some moment instances.

Please update this so it only uses cases where $d was defined as a date.

srijan-paul replied:
Understood. Though, I would've thought using a free function call like `toJSON_(foo)` would be clearer.
I'll change it to call the method instead.

srijan-paul commented on .grit/patterns/moment-to-date-fns.grit:
```diff
@@ -0,0 +1,765 @@
+pattern moment_arith($date, $method, $count, $specifier) {
+  `$date.$method($count, $specifier)` as $arith_exp where {
+    $method <: or { `add` => `addDate`, `subtract` => `subDate` },
+    if ($method <: `add`) {
+      $add = `add`,
+      $add <: ensure_import_from(source=`"date-fns/add"`),
+    } else {
+      $sub = `sub`,
+      $sub <: ensure_import_from(source=`"date-fns/sub"`),
+    }
```
Ah, you're right.
Back when I wrote this method, I had it in mind that the rewrites with `=>` would happen before the `if`, and `$method`'s value would change. I'll refactor it.

srijan-paul commented on .grit/patterns/moment-to-date-fns.grit:
```diff
@@ -0,0 +1,765 @@
+      $exp => `;\n$exp`
+    }
+  }
+}
```
I'll put them in a file called `common.grit`.

srijan-paul commented:
@morgante I've addressed the review comments with the last few commits.
I'm still skeptical about the feasibility of migrating date formatting from moment-js to date-fns.

Earlier, I thought it would be possible to convert a moment.js format specifier to a date-fns compatible one,
but it seems a 1:1 translation is impossible because moment.js's format has features that date-fns dropped in v2 (in order to support standard [unicode tokens](https://www.unicode.org/reports/tr35/tr35-dates.html#Date_Field_Symbol_Table)).
<img width="741" alt="Screenshot 2023-09-30 at 7 36 28 PM" src="https://github.com/getgrit/js/assets/50487716/34da0b6d-910c-44cf-b5d1-709b8624d1df">

For instance, moment.js can parse dates like this:

<img width="632" alt="Screenshot 2023-09-30 at 7 38 20 PM" src="https://github.com/getgrit/js/assets/50487716/0c35cf5b-d6dc-4c3d-becc-417e04df9fc3">

Notice the arbitrary `foobar` that does not map to any Unicode date token.
date-fns v2 simply rejects a format specifier that has `foobar` in it.

In date-fns v1, it is possible:
<img width="637" alt="Screenshot 2023-09-30 at 7 41 34 PM" src="https://github.com/getgrit/js/assets/50487716/2113db93-3a02-4ba0-b716-3c11a19e0c81">

Similarly, moment.js allows interpolation when formatting dates (`"[Today is] DD-MM-YY"`), whereas date-fns v2 dropped support for it.

If we want to support date formatting and parsing the way moment does, we have to migrate to `v1.x` instead of `2.x`.
However, date-fns v1 does not support `Duration`s, which is imperative if we want to support the `Durations` API from moment.js ([ref](https://momentjs.com/docs/#/durations/)), and adding arbitrary amounts of time to a date.

Right now, I see two ways to merge this migration:

1. Migrate to date-fns `v1.x` (wouldn't recommend this) to support `format` and parsing.
But generate a bunch of manual scaffolding to support Durations – an API that doesn't exist in v1.
2. Make this migration _partial_. Whenever formatting/parsing is used, keep both `moment` and `date-fns` import.
Grit would cover 90+% of the migration effort, and the user will need to make very few changes to go the final mile.

IMO, option 2 makes a lot more sense.
We do not support Plugins, global state, and mutating moment.
Hence, an application developer would have to do some manual work anyway.
We can at least migrate them to the modern version of date-fns and add some comments to inform them about the incompatibilities.

If we are to keep the `format` calls in moment library, I can un-draft this PR.

srijan-paul commented on .grit/patterns/moment-to-date-fns.grit:
```diff
@@ -0,0 +1,765 @@
+
+pattern moment_getter_or_setter() {
+  `$d.$accessor($arg)` as $exp where {
+    $accessor <: r"(year|month|week|day|hour|minute|second|millisecond)s?"($unit),
```
I've replaced regexes with `or` wherever feasible.
I think we should still keep this one since there is a necessary capture group

srijan-paul commented on .grit/patterns/moment-to-date-fns.grit:
```diff
@@ -0,0 +1,782 @@
+    `$duration_ctor($count, $literal_unit)` where {
+      // could be moment.duration(...) (default import) or duration(..)
+      $duration_ctor <: or { `$_.duration`, `duration` },
+      // TODO: handle the case where the unit is not a literal (or a compile time constant).
```
```suggestion

```

morgante commented on .grit/patterns/moment-to-date-fns.grit:
```diff
@@ -0,0 +1,781 @@
+    if ($method <: `add`) {
+      $method <: ensure_import_from(source=`"date-fns/add"`),
+    } else {
+      $sub = `sub`,
+      $sub <: ensure_import_from(source=`"date-fns/sub"`),
```
```suggestion
      $method <: ensure_import_from(source=`"date-fns/sub"`),
```

srijan-paul replied:
`$method` here is bound to 'subtract', not `sub`.
moment().subtract() is equivalent to datefns.sub().

morgante commented on .grit/patterns/moment-to-date-fns.grit:
```diff
@@ -0,0 +1,781 @@
+        `relativeTimeThreshold`,
+        `$_.relativeTimeRounding`
+      },
+      $noop = `/* localization in date-fns uses pure functions. ref : https://date-fns.org/v2.30.0/docs/Locale */ void 0`,
```
We should still include the original code in the comment.

```suggestion
      $noop = `/* localization in date-fns uses pure functions.
ref : https://date-fns.org/v2.30.0/docs/Locale
$exp */      
```

srijan-paul commented:
added some negative tests for cases we shouldn't rewrite.

morgante reviewed (approved):
Looks good for now, we can iterate as it is used.

morgante commented:
@srijan-paul Tests aren't passing. Please make sure `grit patterns test` works.

srijan-paul commented:
@morgante Fixed.
Somehow dirtied the file between `grit patterns test` and the commit.

BTW, how do I update the grit-cli locally?
I tried using `pnpm` and `grit install`, but I still seem to get a syntax error (probably because of new features) in `insert_imports` pattern in `imports.grit`.

morgante commented:
> BTW, how do I update the grit-cli locally?

`grit install --update`.
